File: leetcode/0130-surrounded-regions/solution.py
from typing import List


class Solution:
    def solve(self, board: List[List[str]]) -> None:
        if not board or not board[0]:
            return

        rows, cols = len(board), len(board[0])

        def flood(r: int, c: int) -> None:
            stack = [(r, c)]
            board[r][c] = '#'
            while stack:
                cr, cc = stack.pop()
                for nr, nc in ((cr + 1, cc), (cr - 1, cc), (cr, cc + 1), (cr, cc - 1)):
                    if 0 <= nr < rows and 0 <= nc < cols and board[nr][nc] == 'O':
                        board[nr][nc] = '#'
                        stack.append((nr, nc))

        for r in range(rows):
            for c in (0, cols - 1):
                if board[r][c] == 'O':
                    flood(r, c)

        for c in range(cols):
            for r in (0, rows - 1):
                if board[r][c] == 'O':
                    flood(r, c)

        for r in range(rows):
            for c in range(cols):
                if board[r][c] == 'O':
                    board[r][c] = 'X'
                elif board[r][c] == '#':
                    board[r][c] = 'O'

File: leetcode/0130-surrounded-regions/Solution.java
import java.util.ArrayDeque;
import java.util.Deque;

class Solution {
    public void solve(char[][] board) {
        if (board.length == 0 || board[0].length == 0) {
            return;
        }

        int rows = board.length, cols = board[0].length;

        for (int r = 0; r < rows; r++) {
            flood(board, r, 0, rows, cols);
            flood(board, r, cols - 1, rows, cols);
        }
        for (int c = 0; c < cols; c++) {
            flood(board, 0, c, rows, cols);
            flood(board, rows - 1, c, rows, cols);
        }

        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                if (board[r][c] == 'O') {
                    board[r][c] = 'X';
                } else if (board[r][c] == '#') {
                    board[r][c] = 'O';
                }
            }
        }
    }

    private void flood(char[][] board, int r, int c, int rows, int cols) {
        if (board[r][c] != 'O') {
            return;
        }

        Deque<int[]> stack = new ArrayDeque<>();
        stack.push(new int[]{r, c});
        board[r][c] = '#';

        int[][] dirs = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
        while (!stack.isEmpty()) {
            int[] cur = stack.pop();
            for (int[] d : dirs) {
                int nr = cur[0] + d[0], nc = cur[1] + d[1];
                if (nr >= 0 && nr < rows && nc >= 0 && nc < cols && board[nr][nc] == 'O') {
                    board[nr][nc] = '#';
                    stack.push(new int[]{nr, nc});
                }
            }
        }
    }
}
